synapse/cores/ram.py

```diff
@@ -130,8 +132,203 @@ def _getSizeByProp(self, prop, valu=None, mintime=None, maxtime=None):
 
         return len(rows)
 
+
+class SortedCortex(common.Cortex):
+    '''
+    SortedCortex is a RAM-backed Cortex that maintain int prop values
+     in sorted order to facilitate range queries.
+
+    Insertion performance appears to be a constant factor of 10% slower than
+     synapse.cores.ram.Cortex
+    Query by range performance is strictly better.
+    Query by range runtime is O(log n) vs O(n) for the ram:// cortex.
+    '''
+
+    def _initCortex(self):
+
+        def getValu(row):
+            return row[2]
+
+        def makePropSet():
+            return sortedcontainers.SortedSet(key=getValu)
+
+        self.rowsbyid = collections.defaultdict(set)
+        # rows by props with type(value) == string
+        self.srowsbyprop = collections.defaultdict(set)
+        # rows by props with type(value) == int
+        # split because sortedcontainers.SortedSet does not support keys with mixed types
+        self.irowsbyprop = collections.defaultdict(makePropSet)
+        self.rowsbyvalu = collections.defaultdict(set)
+
+        self.initSizeBy('ge',self._sizeByGe)
+        self.initRowsBy('ge',self._rowsByGe)
+
+        self.initSizeBy('le',self._sizeByLe)
+        self.initRowsBy('le',self._rowsByLe)
+
+        self.initSizeBy('range',self._sizeByRange)
+        self.initRowsBy('range',self._rowsByRange)
+
+    def _sizeByRange(self, prop, valu, limit=None):
+        minrow = None
+        if valu[0] != None:
+            if not isint(valu[0]):
+                return 0
+            minrow = (None, None, valu[0], None)
+
+        maxrow = None
+        if valu[1] != None:
+            if not isint(valu[1]):
+                return 0
+            maxrow = (None, None, valu[1], None)
+
+        rowsbyprop = self.irowsbyprop.get(prop)
+        if not rowsbyprop:
+            return 0
+
+        # if minval and maxval are provided, then we are doing a range query,
+        #  of which minval is inclusive, and maxval is exclusive.
+        # otherwise, we only support >= and <=, both of which are inclusive.
+        return sum(1 for _ in rowsbyprop.irange(minimum=minrow, maximum=maxrow, inclusive=(True, not (minrow and maxrow))))
+
+    def _rowsByRange(self, prop, valu, limit=None):
```

Comment:
Seeing 5x instead of the reported 10x, but still a nice bump!
```
Name (time in us)                         Min                 Max               Mean            StdDev             Median               IQR            Outliers(*)  Rounds  Iterations
test_sizeByRange[ram://?sorted=1]      6.5370 (1.0)       97.6550 (1.0)       7.1314 (1.0)      1.9070 (1.0)       6.7400 (1.0)      0.1490 (1.0)        1724;5602   41603           1
test_sizeByRange[ram://]              41.9389 (6.42)     109.4961 (1.12)     43.6742 (6.12)     4.2828 (2.25)     42.5552 (6.31)     0.2645 (1.77)        990;3335   14468           1
```